### PR TITLE
Rework the HIR dump pass

### LINKED ISSUE #693: HIR Dump
Our HIR dump looks like this currently:

```
BlockExpr:
{
 outer attributes: none
 inner attributes: none
 statements: 
 ExprStmtWithoutBlock:
  MethodCallExpr: 
 Object (receiver) expr: ( t ([C: 0 Nid: 97 Hid: 77]))
 Method path segment: 
baz
 Call params:none
 final expression: none
}::[C: 0 Nid: 102 Hid: 81 Lid: 26]
```

Which is very messy and is a clone of our AST as_string methods which is fine for gdb sessions but bad for textural output.

- [ ] Converge on an S-expression format, which is more extensible and easier to read
- [x] Add command line switch https://github.com/Rust-GCC/gccrs/pull/222
- [x] Add dump entry point https://github.com/Rust-GCC/gccrs/blob/28f527c9598339cf834a30b5ee1f14258b8ecbb2/gcc/rust/rust-session-manager.cc#L874-L888
- [x] Dump to file 
- [ ] Print block statements
- [ ] Print block tail expression if present
- [ ] Print crate mapping attributes
- [ ] Print function parameters
- [ ] Visit function's definition block

### PR DESCRIPTION
**Pull request by dkm**
Nearly complete rewrite of the HIR dump pass.
    
fixes #693

### CODE REVIEW

**Comment by dkm** [comment]
Initial work for reworking the HIR dump (easier indentation handling, less formatting work needed,...). Originally motivated by some debugging. Currently simply trying to have something easier to use while keeping the actual output format.
Sample bellow:
```

Crate [
    
    items [
        
        Function {
            func_name: size,
            return_type: usize,
            
            BlockExpr [
                final expression: ( 15 ([C: 0 Nid: 5 Hid: 22]))

            ] // BlockExpr
            node_mappings: [C: 0 Nid: 7 Hid: 24 Lid: 2]

        } // Function
        
        Function {
            func_name: main,
            return_type: void,
            
            BlockExpr [
                
                Stmt [
                    
                    LetStmt {
                        a
                        Expr {
                            
                            ArrayExpr (INCOMPLETE -- CONTENT NOT DISPLAYED) [
                                ArrayExpr:
                                 inner attributes: none

                            ] // ArrayExpr

                        } // Expr

                    } // LetStmt

                ] // Stmt

            ] // BlockExpr
            node_mappings: [C: 0 Nid: 23 Hid: 34 Lid: 3]
...
```

One big improvement is that the dump now clearly shows when it dumps something not yet handled :clown_face:

**Comment by CohenArthur** [comment]
That output looks great :D I like it! I also like the "unimplemented" bit - that's good info to have

**Comment by dkm** [comment]
> That output looks great :D I like it! I also like the "unimplemented" bit - that's good info to have

I still think s-expr would be nicer and easier to parse/browse... But as a first step, I'll try to simply have the content.

File: gcc/rust/hir/tree/rust-hir-expr.h [hunk 1]
```diff
@@ -24,6 +24,7 @@
 #include "rust-hir.h"
 #include "rust-hir-path.h"
 #include "rust-operators.h"
+#include <memory>
```
Comment (philberty) [inline gcc/rust/hir/tree/rust-hir-expr.h:27]: I have a funny feeling this should be included by rust-system.h I think @tschwinge found I was including it and it was marked as poisoned or something in a bootstrap build.

File: gcc/rust/hir/tree/rust-hir-expr.h [hunk 2]
```diff
@@ -1105,8 +1108,8 @@ class ArrayIndexExpr : public ExprWithoutBlock
   void accept_vis (HIRFullVisitor &vis) override;
   void accept_vis (HIRExpressionVisitor &vis) override;
 
-  Expr *get_array_expr () { return array_expr.get (); }
-  Expr *get_index_expr () { return index_expr.get (); }
+  std::unique_ptr<Expr> &get_array_expr () { return array_expr; }
```
Comment (philberty) [inline gcc/rust/hir/tree/rust-hir-expr.h:1111]: You are amazing this has been bugging me that the HIR is super inconsistent like that and returning raw pointers. Thanks for doing this

File: gcc/rust/hir/tree/rust-hir-expr.h [hunk 3]
```diff
@@ -533,8 +539,10 @@ class LazyBooleanExpr : public OperatorExpr
   void accept_vis (HIRFullVisitor &vis) override;
   void accept_vis (HIRExpressionVisitor &vis) override;
 
-  Expr *get_lhs () { return main_or_left_expr.get (); }
+  void visit_lhs (HIRFullVisitor &vis) { main_or_left_expr->accept_vis (vis); }
+  void visit_rhs (HIRFullVisitor &vis) { right_expr->accept_vis (vis); }
 
+  Expr *get_lhs () { return main_or_left_expr.get (); }
```
Comment (philberty) [inline gcc/rust/hir/tree/rust-hir-expr.h:545]: I think this is another case where we should be returning std::unique_ptr

File: gcc/rust/hir/tree/rust-hir-expr.h [hunk 4]
```diff
@@ -459,6 +462,9 @@ class ComparisonExpr : public OperatorExpr
   void accept_vis (HIRFullVisitor &vis) override;
   void accept_vis (HIRExpressionVisitor &vis) override;
 
+  void visit_lhs (HIRFullVisitor &vis) { main_or_left_expr->accept_vis (vis); }
+  void visit_rhs (HIRFullVisitor &vis) { right_expr->accept_vis (vis); }
+
   Expr *get_lhs () { return main_or_left_expr.get (); }
   Expr *get_rhs () { return right_expr.get (); }
```
Comment (philberty) [inline gcc/rust/hir/tree/rust-hir-expr.h:469]: same here

**Review by philberty — COMMENTED** [review]
Great work cant wait to get it merged. Are you still going for s-expressions I have seen dumps in rustc which looked like the AST and another which looked like s-expressions. I don't really care but s-expressions might be useful so we can dump mapping info which would be interesting and eventually extend this pass to dump the type check info too

File: gcc/rust/hir/tree/rust-hir-expr.h (hunk 1 above)
Comment (tschwinge) [reply gcc/rust/hir/tree/rust-hir-expr.h:27]: Yes, instead please (early) `#define INCLUDE_MEMORY`, see <https://github.com/Rust-GCC/gccrs/blob/2a274c1b7bfaa860f41b4bdcfb7008df0bc4cc26/gcc/system.h#L742-L748>.

Comment (dkm) [reply gcc/rust/hir/tree/rust-hir-expr.h:27]: Hmmm, clangd did some nasty stuff behind my back. It automatically adds include files that break the build. I did not add any header file intentionally, so this one is probably the work of clangd.

File: gcc/rust/hir/tree/rust-hir-expr.h (hunk 2 above)
Comment (dkm) [reply gcc/rust/hir/tree/rust-hir-expr.h:1111]: I was not really sure what to do, but I've noticed the inconsistencies. I can fix them along the way, that's easy and helps me understand the HIR anyway :)

File: gcc/rust/hir/tree/rust-hir-expr.h (hunk 3 above)
Comment (dkm) [reply gcc/rust/hir/tree/rust-hir-expr.h:545]: ack

**Comment by dkm** [comment]
> Great work cant wait to get it merged. Are you still going for s-expressions I have seen dumps in rustc which looked like the AST and another which looked like s-expressions. I don't really care but s-expressions might be useful so we can dump mapping info which would be interesting and eventually extend this pass to dump the type check info too

As often for this kind of "this will be quick" work, my initial plan was to keep current format (a rustc-like HIR dump) and have it quickly working. When the first pass is ok-ish, it /should/ not be too hard to move to s-expr. I've tried to use util function with start/stop call, so it should not be that hard to use s-expr. But as a first step, this will be an unspecified rustc-ish HIR dump.

**Comment by dkm** [comment]
So, I think it should be ready to be merged.
Eyeballing the results of some test cases seems ok.
I've executed the full testsuite with the extra `-frust-dump-hir-pretty` and there's no more crash due to assertions or anything else.

I still need to fixup the commit log and add some inline doc/comment, but it should be ready for an initial review (only look at `-dump.{h,cc}` files, the others changes come from #2405 that should go first).

**Comment by dkm** [comment]
I've added few comments, amended the changelog, fixed few minor leftovers... So hopefully this is fine now.

**Comment by dkm** [comment]
Oooops, there are regression after latest rebase, I'll need to investigate before this can be merged.

**Comment by dkm** [comment]
Ok, regressions are fixed, good to go, again.

File: gcc/rust/hir/rust-hir-dump.cc [hunk 5]
```diff
@@ -17,678 +17,2457 @@
...
+    case UseTreeRebind::NewBindType::IDENTIFIER:
+      return "IDENTIFIER";
+    case UseTreeRebind::NewBindType::WILDCARD:
+      return "WILDCARD";
+    }
+  gcc_unreachable ();
+}
+
+std::string
+convert_mut_to_str (Mutability mut)
```
Comment (philberty) [inline gcc/rust/hir/rust-hir-dump.cc:92]: potentially you could move this over to util/rust-common.h i think and leave it as an inline helper

File: gcc/rust/hir/rust-hir-dump.cc [hunk 6]
```diff
@@ -17,678 +17,2457 @@
...
     {
       stream << std::endl;
-      item->accept_vis (*this);
+      beg_of_line = endline;
     }
-  stream << std::endl;
-  stream << indentation;
+}
+
+static std::string delims[2][2] = {
```
Comment (philberty) [inline gcc/rust/hir/rust-hir-dump.cc:150]: i think you should keep this along side the Delim enum as they are associated

**Review by philberty — APPROVED** [review]
LGTM lets get this merged only two minor style things that come to mind.

I think we migt have some changes when we get it merged and start using it but i want to see it first

**Comment by philberty** [comment]
Feel free to merge when your ready

File: gcc/rust/hir/rust-hir-dump.cc (hunk 6 above)
Comment (dkm) [reply gcc/rust/hir/rust-hir-dump.cc:150]: ack

File: gcc/rust/hir/rust-hir-dump.cc (hunk 5 above)
Comment (dkm) [reply gcc/rust/hir/rust-hir-dump.cc]: I've implemented this change, and renamed the function at the same time to `enum_to_str`, a bit more explicit.

**Pull request merged**